### DISCUSSION TITLE
Error when using Cypress to login with Azure Active Directory

### DISCUSSION THREAD

**Discussion by AnimeAllstar** [nextauthjs/next-auth#6888] [Help]
I followed [this tutorial](https://docs.cypress.io/guides/end-to-end-testing/azure-active-directory-authentication) to login via Azure Active Directory using Cypress.

The login works, up until the last point when the following error occurs:

```
[next-auth][error][OAUTH_CALLBACK_ERROR] 
https://next-auth.js.org/errors#oauth_callback_error State cookie was missing. {
  error: TypeError: State cookie was missing.
      at Object.use (/Users/asad/Code/UBC/cpsc319/repos/yojana-next/node_modules/next-auth/core/lib/oauth/checks.js:103:23)
      at oAuthCallback (/Users/asad/Code/UBC/cpsc319/repos/yojana-next/node_modules/next-auth/core/lib/oauth/callback.js:89:25)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async Object.callback (/Users/asad/Code/UBC/cpsc319/repos/yojana-next/node_modules/next-auth/core/routes/callback.js:52:11)
      at async AuthHandler (/Users/asad/Code/UBC/cpsc319/repos/yojana-next/node_modules/next-auth/core/index.js:201:28)
      at async NextAuthHandler (/Users/asad/Code/UBC/cpsc319/repos/yojana-next/node_modules/next-auth/next/index.js:24:19)
      at async /Users/asad/Code/UBC/cpsc319/repos/yojana-next/node_modules/next-auth/next/index.js:60:32
      at async Object.apiResolver (/Users/asad/Code/UBC/cpsc319/repos/yojana-next/node_modules/next/dist/server/api-utils/node.js:372:9)
      at async DevServer.runApi (/Users/asad/Code/UBC/cpsc319/repos/yojana-next/node_modules/next/dist/server/next-server.js:513:9)
      at async Object.fn (/Users/asad/Code/UBC/cpsc319/repos/yojana-next/node_modules/next/dist/server/next-server.js:815:35)
      at async Router.execute (/Users/asad/Code/UBC/cpsc319/repos/yojana-next/node_modules/next/dist/server/router.js:243:32)
      at async DevServer.runImpl (/Users/asad/Code/UBC/cpsc319/repos/yojana-next/node_modules/next/dist/server/base-server.js:432:29)
      at async DevServer.run (/Users/asad/Code/UBC/cpsc319/repos/yojana-next/node_modules/next/dist/server/dev/next-dev-server.js:814:20)
      at async DevServer.handleRequestImpl (/Users/asad/Code/UBC/cpsc319/repos/yojana-next/node_modules/next/dist/server/base-server.js:375:20)
      at async /Users/asad/Code/UBC/cpsc319/repos/yojana-next/node_modules/next/dist/server/base-server.js:157:99 {
    name: 'OAuthCallbackError',
    code: undefined
  },
  providerId: 'azure-ad',
  message: 'State cookie was missing.'
}
```

I am using the [middleware](https://next-auth.js.org/configuration/nextjs#middleware) to protect all my pages and when this error occurs the user is rerouted to the login page.

I am not sure how I am supposed to fix this or if there is an example somewhere with Azure AD + Cypress + Next-Auth.

Any help with how to fix this or an example would help a lot.

**Comment by brandonlenz** [comment]
Not sure if there's a better way yet, but Cypress and Next-auth directly contradict eachother. 

Cypress says never to login via redirects, next-auth has an example encouraging that.

I ran in to a similar issue as you, after allowing Cypress to go through the redirect flow. I think the issue was somehow related to cypress setting cookies differently than when the app is running in a browser normally (this is likely where your issue comes from).

Cypress recommends mocking/programmatically setting the logged in state to your application. I ended up creating a cypress login command that borrowed some code from next-auth's source code to create and store the next auth session cookie. It works, but it's not ideal. 

I would not use Next Auth in a future project in part due to this reason.

**Reply by AnimeAllstar** [reply]
Thanks a lot! Do you mind sharing what you used?

**Reply by brandonlenz** [reply]
@AnimeAllstar it might be a little bit different for you (hopefully less complex), because I have another non-next JS back end that also needs to accept the authentication token sent from browser requests. I created an `/authenticate` endpoint for testing purposes to generate a self signed token from that server, which cypress can then insert into the mocked next auth cookies. 

In my cypress `commands.ts` I added something like the following (edited project specifics out):


```TS
import hkdf from '@panva/hkdf'
import { EncryptJWT, JWTPayload } from 'jose'

// Function derived from https://github.com/nextauthjs/next-auth/blob/d3571e01ba06599ca0411d14d524aa3145ba492b/packages/next-auth/src/jwt/index.ts#L119-L127
async function getDerivedEncryptionKey(secret: string) {
  return await hkdf(
    'sha256',
    secret,
    '',
    'NextAuth.js Generated Encryption Key',
    32
  )
}

// Function derived from https://github.com/nextauthjs/next-auth/blob/d3571e01ba06599ca0411d14d524aa3145ba492b/packages/next-auth/src/jwt/index.ts#L17-L26
async function encode(token: JWTPayload, secret: string) {
  const maxAge = 30 * 24 * 60 * 60 // 30 days
  const encryptionSecret = await getDerivedEncryptionKey(secret)
  return await new EncryptJWT(token)
    .setProtectedHeader({ alg: 'dir', enc: 'A256GCM' })
    .setIssuedAt()
    .setExpirationTime(Math.round(Date.now() / 1000 + maxAge))
    .setJti('test')
    .encrypt(encryptionSecret)
}

Cypress.Commands.add('login', (userObj: JWTPayload) => {
  // Generate and set a valid cookie from the fixture that next-auth can decrypt
  cy.wrap(null)
    .then(() => {
      cy.clearLocalStorage()
      cy.clearCookies()
      return cy
        .request(
          'POST',
          `${Cypress.env('SERVER_BASE_URL')}/authenticate`,
          userObj.sub
        )
        .then((response) => {
          // Set Access Token
          userObj.accessToken = response.body

          // Set Access Token expiration to one hour in future
          userObj.accessTokenExpires = Date.now() + 60 * 60 * 1000

          return encode(userObj, Cypress.env('NEXTAUTH_SECRET'))
        })
    })
    .then((encryptedToken) => {
      const baseUrl = Cypress.config().baseUrl
      const secure = baseUrl.startsWith('https://')
      const cookieName = secure
        ? '__Secure-next-auth.session-token'
        : 'next-auth.session-token'
      cy.setCookie(cookieName, encryptedToken, {
        httpOnly: true,
        secure: true,
      })
    })
})
```

**Reply by brandonlenz** [reply]
@AnimeAllstar haven't tried it yet, but there's some promising newer functionality with [cy.origin](https://docs.cypress.io/api/commands/origin) that may help! 

In my case, I might be able to use https://docs.cypress.io/guides/end-to-end-testing/amazon-cognito-authentication#docusaurus_skipToContent_fallback

**Reply by brandonlenz** [reply]
In your case: https://docs.cypress.io/guides/end-to-end-testing/azure-active-directory-authentication

**Reply by ThreeMangoTrees** [reply]
I followed [the link](https://docs.cypress.io/guides/end-to-end-testing/azure-active-directory-authentication) you added @brandonlenz and I am using `cy.origin()` but still getting the same error as @AnimeAllstar 

Did anyone find any solution to this error?

**Reply by ThreeMangoTrees** [reply]
I have also added the question to stack overflow where I have described the issue in much more detail --> [stack overflow link](https://stackoverflow.com/questions/77754787/state-cookie-missing-error-while-running-cypress-test-to-use-azure-active-direct)

**Reply by murilogtoloni** [reply]
Hey, guys, did you find the solution? I`m having the same issue here... Also, seems that this stack overflow question was removed :(

**Reply by brandonlenz** [reply]
Not helpful, but I no longer use NextAuth (Or NextJS for that matter), or Cypress.

But that's more related to the project I'm currently assigned to. 

That said, this issue still being troublesome would make me avoid NextAuth.

I've entirely switched from cypress to Playwright as well. I haven't used NextAuth in combination with playwright to know if these same issues exist, however.

**Reply by murilogtoloni** [reply]
I found a fix for my scenario! It seems that when the session cookie was set on the initial request it was using the sameSite=Lax policy, so when later the request to exchange the authorization_code with the access token was done I wasn`t getting the session cookie on this request... This is because cypress runs our projects inside an Iframe which is not a first level layer of browsing, so sameSite=Lax prevents the request of having these cookies.

So the solution was to add a intercept to the initial call that is made to mutate the Set-Cookie header by changing the sameSite=Lax to sameSite=None

something like this(you can change to your default auth endpoint that sets the session cookie):

`cy.intercept('*', (req) => {
    req.on('response', (res) => {
      const setCookies = res.headers['set-cookie'];
      if (setCookies) {
        res.headers['set-cookie'] = (
          Array.isArray(setCookies) ? setCookies : [setCookies]
        )
          .filter(Boolean)
          .map((headerContent) =>
            headerContent.replace(
              /samesite=(lax|strict)/gi,
              'SameSite=None; Secure'
            )
          );
      }
    });
  });`

**Reply by austinb847** [reply]
@murilotoloni you're a life saver! it also fixes the issue if you adjust the `checks` property in the provider config to this: 

```
 MicrosoftEntraID({
      ...other configs,
      checks: ['none']
    }),
```

but I much prefer your solution for security purposes!
